Route binding : CF-ServiceBrokerBadResponse(10001) due to 502 Bad Gateway

While testing the binding of a route, we get a CF-ServiceBrokerBadResponse(10001) whenever trying to send a response.

```
2016-04-18T15:49:39.35+0200 [APP/0]      OUT Caused by: org.cloudfoundry.client.v2.CloudFoundryException: CF-ServiceBrokerBadResponse(10001): The service broker returned an invalid response for the request to http://autosleep-arno.cf.ns.nd-paas.itn.ftgroup/v2/service_instances/141f0d49-59f5-493f-afb6-086b427b0730/service_bindings/b29d5495-e4b5-404c-986e-834d092b4b58. Status Code: 201 Created, Body: {"route_service_url":"autosleep-arno.cf.ns.nd-paas.itn.ftgroup/proxy/b29d5495-e4b5-404c-986e-834d092b4b58"}
2016-04-18T15:49:39.35+0200 [APP/0]      OUT    at org.cloudfoundry.spring.client.v2.CloudFoundryExceptionBuilder.build(CloudFoundryExceptionBuilder.java:47) ~[cloudfoundry-client-spring-2.0.0.M1.jar!/:2.0.0.M1]
2016-04-18T15:49:39.35+0200 [APP/0]      OUT    at org.cloudfoundry.spring.util.AbstractSpringOperations.lambda$null$1(AbstractSpringOperations.java:72) ~[cloudfoundry-client-spring-2.0.0.M1.jar!/:2.0.0.M1]
2016-04-18T15:49:39.35+0200 [APP/0]      OUT    at reactor.core.publisher.FluxYieldingEmitter$YieldingSignalEmitter.request(FluxYieldingEmitter.java:72) ~[reactor-core-2.5.0.M2.jar!/:na]
2016-04-18T15:49:39.35+0200 [APP/0]      OUT    at reactor.core.util.DeferredSubscription.request(DeferredSubscription.java:92) ~[reactor-core-2.5.0.M2.jar!/:na]
2016-04-18T15:49:39.35+0200 [APP/0]      OUT    at reactor.core.publisher.FluxPublishOn$PublishOnPipeline.run(FluxPublishOn.java:164) ~[reactor-core-2.5.0.M2.jar!/:na]
2016-04-18T15:49:39.35+0200 [APP/0]      OUT    at reactor.core.publisher.SchedulerGroup$TaskSubscriber.onNext(SchedulerGroup.java:1081) ~[reactor-core-2.5.0.M2.jar!/:na]
2016-04-18T15:49:39.35+0200 [APP/0]      OUT    at reactor.core.publisher.SchedulerGroup$TaskSubscriber.onNext(SchedulerGroup.java:1063) ~[reactor-core-2.5.0.M2.jar!/:na]
2016-04-18T15:49:39.35+0200 [APP/0]      OUT    at reactor.core.publisher.WorkQueueProcessor$QueueSubscriberLoop.run(WorkQueueProcessor.java:842) ~[reactor-core-2.5.0.M2.jar!/:na]
2016-04-18T15:49:39.35+0200 [APP/0]      OUT    ... 3 common frames omitted
2016-04-18T15:49:39.35+0200 [APP/0]      OUT Caused by: org.springframework.web.client.HttpServerErrorException: 502 Bad Gateway
2016-04-18T15:49:39.35+0200 [APP/0]      OUT    at org.springframework.web.client.DefaultResponseErrorHandler.handleError(DefaultResponseErrorHandler.java:94) ~[spring-web-4.2.5.RELEASE.jar!/:4.2.5.RELEASE]
2016-04-18T15:49:39.35+0200 [APP/0]      OUT    at org.springframework.security.oauth2.client.http.OAuth2ErrorHandler.handleError(OAuth2ErrorHandler.java:85) ~[spring-security-oauth2-2.0.9.RELEASE.jar!/:na]
2016-04-18T15:49:39.35+0200 [APP/0]      OUT    at org.springframework.web.client.RestTemplate.handleResponse(RestTemplate.java:641) ~[spring-web-4.2.5.RELEASE.jar!/:4.2.5.RELEASE]
2016-04-18T15:49:39.35+0200 [APP/0]      OUT    at org.springframework.web.client.RestTemplate.doExecute(RestTemplate.java:597) ~[spring-web-4.2.5.RELEASE.jar!/:4.2.5.RELEASE]
2016-04-18T15:49:39.35+0200 [APP/0]      OUT    at org.springframework.security.oauth2.client.OAuth2RestTemplate.doExecute(OAuth2RestTemplate.java:128) ~[spring-security-oauth2-2.0.9.RELEASE.jar!/:na]
2016-04-18T15:49:39.35+0200 [APP/0]      OUT    at org.springframework.web.client.RestTemplate.execute(RestTemplate.java:572) ~[spring-web-4.2.5.RELEASE.jar!/:4.2.5.RELEASE]
2016-04-18T15:49:39.35+0200 [APP/0]      OUT    at org.springframework.web.client.RestTemplate.exchange(RestTemplate.java:534) ~[spring-web-4.2.5.RELEASE.jar!/:4.2.5.RELEASE]
2016-04-18T15:49:39.35+0200 [APP/0]      OUT    at org.cloudfoundry.spring.util.AbstractSpringOperations.lambda$putWithBody$10(AbstractSpringOperations.java:139) ~[cloudfoundry-client-spring-2.0.0.M1.jar!/:2.0.0.M1]
2016-04-18T15:49:39.35+0200 [APP/0]      OUT    at org.cloudfoundry.spring.util.AbstractSpringOperations.lambda$null$1(AbstractSpringOperations.java:66) ~[cloudfoundry-client-spring-2.0.0.M1.jar!/:2.0.0.M1]
2016-04-18T15:49:39.35+0200 [APP/0]      OUT    ... 9 common frames omitted

```

More complete log:

```
2016-04-18T15:54:42.40+0200 [APP/0]      OUT ERROR o.c.a.w.ApplicationStopper(128) - error while requesting remote api
2016-04-18T15:54:42.40+0200 [APP/0]      OUT org.cloudfoundry.autosleep.worker.remote.CloudFoundryException: org.cloudfoundry.client.v2.CloudFoundryException: CF-ServiceBrokerBadResponse(10001): The service broker returned an invalid response for the request to http://autosleep-arno.cf.ns.nd-paas.itn.ftgroup/v2/service_instances/141f0d49-59f5-493f-afb6-086b427b0730/service_bindings/3b23dd53-7af8-4918-9f27-78f3231f53f6. Status Code: 201 Created, Body: {"route_service_url":"autosleep-arno.cf.ns.nd-paas.itn.ftgroup/proxy/3b23dd53-7af8-4918-9f27-78f3231f53f6"}
2016-04-18T15:54:42.40+0200 [APP/0]      OUT    at org.cloudfoundry.autosleep.worker.remote.CloudFoundryApi.waitForResult(CloudFoundryApi.java:361) ~[classes/:na]
2016-04-18T15:54:42.40+0200 [APP/0]      OUT    at org.cloudfoundry.autosleep.worker.remote.CloudFoundryApi.bind(CloudFoundryApi.java:124) ~[classes/:na]
2016-04-18T15:54:42.40+0200 [APP/0]      OUT    at org.cloudfoundry.autosleep.worker.remote.CloudFoundryApi.bindRoutes(CloudFoundryApi.java:141) ~[classes/:na]
2016-04-18T15:54:42.40+0200 [APP/0]      OUT    at org.cloudfoundry.autosleep.worker.ApplicationStopper.putApplicationToSleep(ApplicationStopper.java:158) [classes/:na]
2016-04-18T15:54:42.40+0200 [APP/0]      OUT    at org.cloudfoundry.autosleep.worker.ApplicationStopper.checkActiveApplication(ApplicationStopper.java:88) [classes/:na]
2016-04-18T15:54:42.40+0200 [APP/0]      OUT    at org.cloudfoundry.autosleep.worker.ApplicationStopper.handleApplicationEnrolled(ApplicationStopper.java:125) [classes/:na]
2016-04-18T15:54:42.40+0200 [APP/0]      OUT    at org.cloudfoundry.autosleep.worker.ApplicationStopper.lambda$run$19(ApplicationStopper.java:179) [classes/:na]
2016-04-18T15:54:42.40+0200 [APP/0]      OUT    at org.cloudfoundry.autosleep.util.ApplicationLocker.executeThreadSafe(ApplicationLocker.java:47) ~[classes/:na]
2016-04-18T15:54:42.40+0200 [APP/0]      OUT    at org.cloudfoundry.autosleep.worker.ApplicationStopper.run(ApplicationStopper.java:172) [classes/:na]
2016-04-18T15:54:42.40+0200 [APP/0]      OUT    at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) ~[na:1.8.0_65-]
2016-04-18T15:54:42.40+0200 [APP/0]      OUT    at java.util.concurrent.FutureTask.run(FutureTask.java:266) ~[na:1.8.0_65-]
2016-04-18T15:54:42.40+0200 [APP/0]      OUT    at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180) ~[na:1.8.0_65-]
2016-04-18T15:54:42.40+0200 [APP/0]      OUT    at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293) ~[na:1.8.0_65-]
2016-04-18T15:54:42.40+0200 [APP/0]      OUT    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) ~[na:1.8.0_65-]
2016-04-18T15:54:42.40+0200 [APP/0]      OUT    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) ~[na:1.8.0_65-]
2016-04-18T15:54:42.40+0200 [APP/0]      OUT    at java.lang.Thread.run(Thread.java:745) ~[na:1.8.0_65-]
2016-04-18T15:54:42.40+0200 [APP/0]      OUT Caused by: org.cloudfoundry.client.v2.CloudFoundryException: CF-ServiceBrokerBadResponse(10001): The service broker returned an invalid response for the request to http://autosleep-arno.cf.ns.nd-paas.itn.ftgroup/v2/service_instances/141f0d49-59f5-493f-afb6-086b427b0730/service_bindings/3b23dd53-7af8-4918-9f27-78f3231f53f6. Status Code: 201 Created, Body: {"route_service_url":"autosleep-arno.cf.ns.nd-paas.itn.ftgroup/proxy/3b23dd53-7af8-4918-9f27-78f3231f53f6"}
2016-04-18T15:54:42.40+0200 [APP/0]      OUT    at org.cloudfoundry.spring.client.v2.CloudFoundryExceptionBuilder.build(CloudFoundryExceptionBuilder.java:47) ~[cloudfoundry-client-spring-2.0.0.M1.jar!/:2.0.0.M1]
2016-04-18T15:54:42.40+0200 [APP/0]      OUT    at org.cloudfoundry.spring.util.AbstractSpringOperations.lambda$null$1(AbstractSpringOperations.java:72) ~[cloudfoundry-client-spring-2.0.0.M1.jar!/:2.0.0.M1]
2016-04-18T15:54:42.40+0200 [APP/0]      OUT    at reactor.core.publisher.FluxYieldingEmitter$YieldingSignalEmitter.request(FluxYieldingEmitter.java:72) ~[reactor-core-2.5.0.M2.jar!/:na]
2016-04-18T15:54:42.40+0200 [APP/0]      OUT    at reactor.core.util.DeferredSubscription.set(DeferredSubscription.java:71) ~[reactor-core-2.5.0.M2.jar!/:na]
2016-04-18T15:54:42.40+0200 [APP/0]      OUT    at reactor.core.publisher.FluxPublishOn$PublishOnPipeline.onSubscribe(FluxPublishOn.java:126) ~[reactor-core-2.5.0.M2.jar!/:na]
2016-04-18T15:54:42.40+0200 [APP/0]      OUT    at reactor.core.subscriber.SignalEmitter.start(SignalEmitter.java:170) ~[reactor-core-2.5.0.M2.jar!/:na]
2016-04-18T15:54:42.40+0200 [APP/0]      OUT    at reactor.core.publisher.FluxYieldingEmitter.subscribe(FluxYieldingEmitter.java:43) ~[reactor-core-2.5.0.M2.jar!/:na]
2016-04-18T15:54:42.40+0200 [APP/0]      OUT    at reactor.core.publisher.FluxFlatMap.trySubscribeScalarMap(FluxFlatMap.java:152) ~[reactor-core-2.5.0.M2.jar!/:na]
2016-04-18T15:54:42.40+0200 [APP/0]      OUT    at reactor.core.publisher.FluxFlatMap.subscribe(FluxFlatMap.java:169) ~[reactor-core-2.5.0.M2.jar!/:na]
2016-04-18T15:54:42.40+0200 [APP/0]      OUT    at reactor.core.publisher.FluxPublishOn$SourceSubscribeTask.run(FluxPublishOn.java:308) ~[reactor-core-2.5.0.M2.jar!/:na]
2016-04-18T15:54:42.40+0200 [APP/0]      OUT    at reactor.core.publisher.SchedulerGroup$TaskSubscriber.onNext(SchedulerGroup.java:1081) ~[reactor-core-2.5.0.M2.jar!/:na]
2016-04-18T15:54:42.40+0200 [APP/0]      OUT    at reactor.core.publisher.SchedulerGroup$TaskSubscriber.onNext(SchedulerGroup.java:1063) ~[reactor-core-2.5.0.M2.jar!/:na]
2016-04-18T15:54:42.40+0200 [APP/0]      OUT    at reactor.core.publisher.WorkQueueProcessor$QueueSubscriberLoop.run(WorkQueueProcessor.java:842) ~[reactor-core-2.5.0.M2.jar!/:na]
2016-04-18T15:54:42.40+0200 [APP/0]      OUT    ... 3 common frames omitted
2016-04-18T15:54:42.40+0200 [APP/0]      OUT Caused by: org.springframework.web.client.HttpServerErrorException: 502 Bad Gateway
2016-04-18T15:54:42.40+0200 [APP/0]      OUT    at org.springframework.web.client.DefaultResponseErrorHandler.handleError(DefaultResponseErrorHandler.java:94) ~[spring-web-4.2.5.RELEASE.jar!/:4.2.5.RELEASE]
2016-04-18T15:54:42.40+0200 [APP/0]      OUT    at org.springframework.security.oauth2.client.http.OAuth2ErrorHandler.handleError(OAuth2ErrorHandler.java:85) ~[spring-security-oauth2-2.0.9.RELEASE.jar!/:na]
2016-04-18T15:54:42.40+0200 [APP/0]      OUT    at org.springframework.web.client.RestTemplate.handleResponse(RestTemplate.java:641) ~[spring-web-4.2.5.RELEASE.jar!/:4.2.5.RELEASE]
2016-04-18T15:54:42.40+0200 [APP/0]      OUT    at org.springframework.web.client.RestTemplate.doExecute(RestTemplate.java:597) ~[spring-web-4.2.5.RELEASE.jar!/:4.2.5.RELEASE]
2016-04-18T15:54:42.40+0200 [APP/0]      OUT    at org.springframework.security.oauth2.client.OAuth2RestTemplate.doExecute(OAuth2RestTemplate.java:128) ~[spring-security-oauth2-2.0.9.RELEASE.jar!/:na]
2016-04-18T15:54:42.40+0200 [APP/0]      OUT    at org.springframework.web.client.RestTemplate.execute(RestTemplate.java:572) ~[spring-web-4.2.5.RELEASE.jar!/:4.2.5.RELEASE]
2016-04-18T15:54:42.40+0200 [APP/0]      OUT    at org.springframework.web.client.RestTemplate.exchange(RestTemplate.java:534) ~[spring-web-4.2.5.RELEASE.jar!/:4.2.5.RELEASE]
2016-04-18T15:54:42.40+0200 [APP/0]      OUT    at org.cloudfoundry.spring.util.AbstractSpringOperations.lambda$putWithBody$10(AbstractSpringOperations.java:139) ~[cloudfoundry-client-spring-2.0.0.M1.jar!/:2.0.0.M1]
2016-04-18T15:54:42.40+0200 [APP/0]      OUT    at org.cloudfoundry.spring.util.AbstractSpringOperations.lambda$null$1(AbstractSpringOperations.java:66) ~[cloudfoundry-client-spring-2.0.0.M1.jar!/:2.0.0.M1]
2016-04-18T15:54:42.40+0200 [APP/0]      OUT    ... 14 common frames omitted
```

--- a/src/main/java/org/cloudfoundry/autosleep/ui/servicebroker/configuration/AutosleepCatalogBuilder.java
+++ b/src/main/java/org/cloudfoundry/autosleep/ui/servicebroker/configuration/AutosleepCatalogBuilder.java
@@ -24,6 +24,7 @@ import org.springframework.cloud.servicebroker.model.Catalog;
 import org.springframework.cloud.servicebroker.model.Plan;
 import org.springframework.cloud.servicebroker.model.ServiceDefinition;
 import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.cloud.servicebroker.model.ServiceDefinitionRequires;
 import org.springframework.context.annotation.Bean;
 import org.springframework.context.annotation.Configuration;
 import org.springframework.core.env.Environment;
@@ -61,7 +62,7 @@ public class AutosleepCatalogBuilder {
                                 true)),
                 Arrays.asList("autosleep", "document"),
                 getServiceDefinitionMetadata(),
-                Collections.singletonList("route_forwarding"),
+                Collections.singletonList(ServiceDefinitionRequires.SERVICE_REQUIRES_ROUTE_FORWARDING.toString()),
                 null)));
     }
 

--- a/src/main/java/org/cloudfoundry/autosleep/ui/servicebroker/service/AutosleepBindingService.java
+++ b/src/main/java/org/cloudfoundry/autosleep/ui/servicebroker/service/AutosleepBindingService.java
@@ -88,7 +88,7 @@ public class AutosleepBindingService implements ServiceInstanceBindingService {
 
         String targetAppId = (String) request.getBindResource()
                 .get(ServiceBindingResource.BIND_RESOURCE_KEY_APP.toString());
-        String routeId = (String) request.getBindResource()
+        String route = (String) request.getBindResource()
                 .get(ServiceBindingResource.BIND_RESOURCE_KEY_ROUTE.toString());
 
         Binding.BindingBuilder bindingBuilder = Binding.builder()
@@ -116,23 +116,10 @@ public class AutosleepBindingService implements ServiceInstanceBindingService {
             });
             return new CreateServiceInstanceAppBindingResponse().withCredentials(Collections.singletonMap(
                     Config.ServiceInstanceParameters.IDLE_DURATION, spaceEnrollerConfig.getIdleDuration().toString()));
-        } else if (routeId != null) {
-            log.debug("creating binding {} for route {}", bindingId, routeId);
+        } else if (route != null) {
+            log.debug("creating binding {} for route {}", bindingId, route);
 
-            //check we know at least one app or all?
-            try {
-                List<String> appIds = cfApi.listRouteApplications(routeId);
-                log.debug("CF knows {} apps, amoung which autosleep knows {}", appIds.size(),
-                        appRepository.countByApplicationIds(appIds));
-                if (appRepository.countByApplicationIds(appIds) != appIds.size()) {
-                    throw new ServiceBrokerException("Only Autosleep is allowed to bind route to itself");
-                }
-            } catch (CloudFoundryException e) {
-                throw new ServiceBrokerException("Can't check applications map to this route id");
-            }
-
-            bindingBuilder.resourceId(routeId)
-                    .resourceType(Route);
+            bindingBuilder.resourceId(route).resourceType(Route);
 
             bindingRepository.save(bindingBuilder.build());
 
@@ -142,7 +129,7 @@ public class AutosleepBindingService implements ServiceInstanceBindingService {
             }
 
             return new CreateServiceInstanceRouteBindingResponse()
-                    .withRouteServiceUrl(firstUri + Config.Path.PROXY_CONTEXT + "/" + bindingId);
+                    .withRouteServiceUrl("https://" + firstUri + Config.Path.PROXY_CONTEXT + "/" + bindingId);
         } else {
             throw new ServiceBrokerException("Unknown bind resource type");
         }

--- a/src/test/java/org/cloudfoundry/autosleep/ui/servicebroker/service/AutosleepBindingServiceTest.java
+++ b/src/test/java/org/cloudfoundry/autosleep/ui/servicebroker/service/AutosleepBindingServiceTest.java
@@ -294,38 +294,6 @@ public class AutosleepBindingServiceTest {
     }
 
     @Test
-    public void new_route_binding_should_scream_cf_api_fails() throws Exception {
-        //given that the app is linked to the route in CF
-        when(cfApi.listRouteApplications(ROUTE_UID)).thenThrow(new CloudFoundryException(new Exception()));
-
-        //when receive a new ROUTE binding
-        //then it should scream, and not save anything
-        verifyThrown(() -> bindingService.createServiceInstanceBinding(createRouteBindingTemplate
-                .withServiceInstanceId("Sid")
-                .withBindingId("Bidroute")), ServiceBrokerException.class);
-        verify(bindingRepository, never()).save(any(Binding.class));
-    }
-
-    /**
-     * Test that only autosleep can call itself on a route binding (no manual route binding).
-     */
-    @Test
-    public void new_route_binding_should_scream_if_unknown_appid() throws Exception {
-        //given that the app is linked to the route in CF
-        when(cfApi.listRouteApplications(ROUTE_UID)).thenReturn(singletonList(APP_UID));
-        //given that the application is unknown from autosleep
-        when(appRepo.findOne(APP_UID)).thenReturn(null);
-
-        //when receive a new ROUTE binding
-
-        //then it should scream, and not save anything
-        verifyThrown(() -> bindingService.createServiceInstanceBinding(createRouteBindingTemplate
-                .withServiceInstanceId("Sid")
-                .withBindingId("Bidroute")), ServiceBrokerException.class);
-        verify(bindingRepository, never()).save(any(Binding.class));
-    }
-
-    @Test
     public void new_route_binding_should_store_binding() throws Exception {
         //given that the app is linked to the route in CF
         when(cfApi.listRouteApplications(ROUTE_UID)).thenReturn(singletonList(APP_UID));

